service/pkg/config/environment_value_loader.go

```diff
@@ -0,0 +1,135 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/spf13/viper"
+)
+
+// EnvironmentValueLoader implements Loader using Viper
+type EnvironmentValueLoader struct {
+	allowListMap map[string]struct{}
+	viper        *viper.Viper
+}
+
+// NewEnvironmentValueLoader creates a new Viper-based configuration loader
+// to load from environment variables, from a default or specified file
+// (or k8s config map), or some combination
+func NewEnvironmentValueLoader(key string, allowList []string) (*EnvironmentValueLoader, error) {
+	// Set paths and config file info
+	v := viper.NewWithOptions(viper.WithLogger(slog.Default()))
+
+	// Environment variable settings
+	v.SetEnvPrefix(key)
+	v.SetEnvKeyReplacer(strings.NewReplacer(".", "_"))
+	v.AutomaticEnv()
+
+	var allowListMap map[string]struct{}
+	if allowList != nil || len(allowList) > 0 {
+		allowListMap = make(map[string]struct{})
+		for _, allow := range allowList {
+			allowListMap[allow] = struct{}{}
+		}
+	}
+
+	result := &EnvironmentValueLoader{
+		allowListMap: allowListMap,
+		viper:        v,
+	}
+	return result, nil
+}
+
+// Get fetches a particular config value by dot-delimited key from the source
+func (l *EnvironmentValueLoader) Get(key string) (any, error) {
+	if l.allowListMap != nil {
+		if _, keyInAllowList := l.allowListMap[key]; !keyInAllowList {
+			return nil, fmt.Errorf("environment value %s is not allowed", key)
+		}
+	}
+	return l.viper.Get(key), nil
+}
+
+// GetConfigKeys returns all the configuration keys found in the environment variables.
+func (l *EnvironmentValueLoader) GetConfigKeys() ([]string, error) {
+	return l.viper.AllKeys(), nil
+}
+
+// Load loads the configuration into the provided struct
+func (l *EnvironmentValueLoader) Load(_ Config) error {
+	// For environment variables, Viper's `AutomaticEnv` handles this, so no explicit load is needed here.
+	return nil
+}
+
+// Watch starts watching the config file for configuration changes
+func (l *EnvironmentValueLoader) Watch(ctx context.Context, _ *Config, onChange func(context.Context) error) error {
+	// Environment variables can't be watched directly, so we poll them.
+	interval := defaultWatchViaPollInterval
+	slog.DebugContext(ctx, "starting environment configuration polling", slog.Duration("interval", interval))
+
+	ticker := time.NewTicker(interval)
+
+	go func() {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ticker.C:
+				slog.DebugContext(ctx, "polling environment variables for changes")
```

Comment:
Does this work when tested? I think I tested this in the past and found the running Go process could not successfully poll for environment variable changes at runtime.

Reply:
Good question. I actually don't think I did test that. And true, I'm not sure how I even _would_ change the running env to test without a manual setenv somewhere. Happy to test though, let me see what I can dig up for testing.

Reply:
Ripped it out!